src/btree/bt_cursor.c

```diff
@@ -971,19 +971,24 @@ static int
 __curfile_update_check(WT_CURSOR_BTREE *cbt)
 {
     WT_BTREE *btree;
+    WT_PAGE *page;
     WT_SESSION_IMPL *session;
 
     btree = cbt->btree;
+    page = cbt->ref->page;
     session = (WT_SESSION_IMPL *)cbt->iface.session;
 
     if (cbt->compare != 0)
         return (0);
     if (cbt->ins != NULL)
-        return (__wt_txn_update_check(session, cbt->ins->upd));
+        return (__wt_txn_update_check(session, cbt, cbt->ins->upd));
+
+    if (btree->type == BTREE_ROW && page->modify != NULL && page->modify->mod_row_update != NULL)
+        return (__wt_txn_update_check(session, cbt, page->modify->mod_row_update[cbt->slot]));
+
+    if (btree->type == BTREE_COL_VAR)
+        return (__wt_txn_update_check(session, cbt, NULL));
```

Comment:
Can the above logic be folded into __wt_txn_update_check as well? I think it'll be better if all the necessary decision making about whether to check an on-page value is in one place. Once we have that it'll be easier to reason about what is going wrong in the failure case.